bootstrap/utils.go

```diff
@@ -0,0 +1,378 @@
+package bootstrap
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/square/go-jose.v2"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"stash.kopano.io/kc/konnect/signing"
+	"strings"
+)
+
+func parseJSONWebKey(jsonBytes []byte) (*jose.JSONWebKey, error) {
+	k := &jose.JSONWebKey{}
+	if err := k.UnmarshalJSON(jsonBytes); err != nil {
+		return nil, err
+	}
+	return k, nil
+}
+
+func LoadSignerFromFile(fn string) (string, crypto.Signer, error) {
+	readBytes, errRead := ioutil.ReadFile(fn)
+	if errRead != nil {
+		return "", nil, fmt.Errorf("failed to parse key file: %v", errRead)
+	}
+
+	ext := filepath.Ext(fn)
+	switch ext {
+	case ".json":
+		k, err := parseJSONWebKey(readBytes)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to parse key file as JWK: %v", err)
+		}
+		if !k.Valid() {
+			return "", nil, fmt.Errorf("json file is not a valid JWK")
+		}
+		if k.IsPublic() {
+			return "", nil, fmt.Errorf("JWK is a public key, private key required to use as signer")
+		}
+		signer, ok := k.Key.(crypto.Signer)
+		if !ok {
+			return "", nil, fmt.Errorf("JWS key type %T is not a signer", k.Key)
+		}
+
+		return k.KeyID, signer, nil
+
+	case ".pem":
+		fallthrough
+	default:
+		// Try PEM if not otherwise detected.
+		signer, err := parsePEMSigner(readBytes)
+		return "", signer, err
+	}
+}
+
+func parsePEMSigner(pemBytes []byte) (crypto.Signer, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, fmt.Errorf("no PEM block found")
+	}
+
+	var signer crypto.Signer
+	for {
+		pkcs1Key, errParse1 := x509.ParsePKCS1PrivateKey(block.Bytes)
+		if errParse1 == nil {
+			signer = pkcs1Key
+			break
+		}
+
+		pkcs8Key, errParse2 := x509.ParsePKCS8PrivateKey(block.Bytes)
+		if errParse2 == nil {
+			signerSigner, ok := pkcs8Key.(crypto.Signer)
+			if !ok {
+				return nil, fmt.Errorf("failed to use key as crypto signer")
+			}
+			signer = signerSigner
+			break
+		}
+
+		ecKey, errParse3 := x509.ParseECPrivateKey(block.Bytes)
+		if errParse3 == nil {
+			signer = ecKey
+			break
+		}
+
+		return nil, fmt.Errorf("failed to parse signer key - valid PKCS#1, PKCS#8 ...? %v, %v, %v", errParse1, errParse2, errParse3)
+	}
+
+	return signer, nil
+}
+
+func LoadValidatorFromFile(fn string) (string, crypto.PublicKey, error) {
```

Comment:
Same here, comment missing if needs to be exported.

Reply:
There are more, in this file with the same issue - i will skip commenting each individually :)